sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -1170,25 +1233,56 @@ reduAuxCGFunc(handler &CGH, size_t NWorkItems, size_t MaxWGSize,
   // Get read accessor to the buffer that was used as output
   // in the previous kernel.
   auto In = Redu.getReadAccToPreviousPartialReds(CGH);
-  if (Reduction::is_usm && NWorkGroups == 1) {
-    if (HasUniformWG)
-      reduAuxCGFuncImpl<KernelName, KernelType, true>(
-          CGH, NWorkItems, NWorkGroups, WGSize, Redu, In, Redu.getUSMPointer());
-    else
-      reduAuxCGFuncImpl<KernelName, KernelType, false>(
-          CGH, NWorkItems, NWorkGroups, WGSize, Redu, In, Redu.getUSMPointer());
-  } else {
-    auto Out = Redu.getWriteAccForPartialReds(NWorkGroups, CGH);
-    if (HasUniformWG)
-      reduAuxCGFuncImpl<KernelName, KernelType, true>(
-          CGH, NWorkItems, NWorkGroups, WGSize, Redu, In, Out);
-    else
-      reduAuxCGFuncImpl<KernelName, KernelType, false>(
-          CGH, NWorkItems, NWorkGroups, WGSize, Redu, In, Out);
-  }
+  auto Out = Redu.getWriteAccForPartialReds(NWorkGroups, CGH);
+  if (HasUniformWG)
+    reduAuxCGFuncImpl<KernelName, KernelType, true>(
+        CGH, NWorkItems, NWorkGroups, WGSize, Redu, In, Out);
+  else
+    reduAuxCGFuncImpl<KernelName, KernelType, false>(
+        CGH, NWorkItems, NWorkGroups, WGSize, Redu, In, Out);
   return NWorkGroups;
 }
 
+// This method is used for implementation of parallel_for accepting 1 reduction.
+// TODO: remove this method when everything is switched to general algorithm
+// implementing arbitrary number of reductions in parallel_for().
+/// Copies the final reduction result kept in read-write accessor to user's
+/// accessor. This method is not called for user's read-write accessors
+/// requiring update-write to it.
+template <typename KernelName, class Reduction>
+std::enable_if_t<!Reduction::is_usm>
+reduSaveFinalResultToUserMem(handler &CGH, Reduction &Redu) {
+  auto InAcc = Redu.getReadAccToPreviousPartialReds(CGH);
+#ifndef __SYCL_DEVICE_ONLY__
+  Redu.associateWithHandler(CGH);
+#endif
```

Comment:
Why is this guarded by `__SYCL_DEVICE_ONLY__`?  Since `CGH.copy` is called below, this function can't ever be executed on the device, can it?

Reply:
Right. I'll remove #ifdefs here. At this place there is no need in guarding this Redu.associateWithHandler(CGH) line because all required guards already available inside that method. The latter is needed because the method handler::associateWithHandler() is guarded (i.e. that method is not defined when __SYCL_DEVICE_ONLY___)